### PR TITLE
Update libsecp256k1, Go wrapper and tests

### PR DESCRIPTION
**Pull request by Gustav-Simonsson**
**NOTE:** ```RecoverPubkey``` is on the consensus-critical code path. Please review carefully.

Go wrapper update is in it's own commit, files changed:

```
crypto/secp256k1/secp256.go
crypto/secp256k1/secp256_test.go
```

```
benchmark                                  old ns/op     new ns/op     delta
BenchmarkSignRandomInputEachRound-2        148648        149994        +0.91%
BenchmarkRecoverRandomInputEachRound-2     207831        203833        -1.92%
```

We'll probably see a more significant speedup when we can improve internal conversions (as in remove where possible) of signatures and pubkeys - currently we need two extra calls to the lib for each recovery - one to serialise signature and one to serialise the returned pubkey.

If we can avoid those calls + other internal conversions we could probably gain a few % more on recovery.

- [x] Update libsecp256k1
- [x] Update Go wrapper to new libsecp256k1 API
- [x] Refactor Go wrapper tests to have meaningful names, clearer validations and a few more edge cases.
- [x] Add benchmarks of ```Sign``` and ```RecoverPubkey```

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Mon Sep 28 15:43:22 UTC 2015_

**Comment by fjl** [comment]
Did you add zeroing of intermediate buffers yet?

**Comment by Gustav-Simonsson** [comment]
@fjl not yet.

**Comment by obscuren** [comment]
LGTM :+1:

**Comment by Gustav-Simonsson** [comment]
@obscuren removed unneeded return matchings. `SECP256K1_EC_COMPRESSED` is from the name of the macro in the lib

File: crypto/secp256k1/secp256_test.go [hunk 1]
```diff
@@ -188,67 +149,85 @@ func randSig() []byte {
...
+//godep go test -v -run=XXX -bench=BenchmarkECRec
+func BenchmarkRecover(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		_, seckey := GenerateKeyPair()
+		msg := randentropy.GetEntropyCSPRNG(32)
+		sig, _ := Sign(msg, seckey)
+		b.StartTimer()
+		_, e := RecoverPubkey(msg, sig)
+		err = e
+		b.StopTimer()
```
Comment (fjl) [inline crypto/secp256k1/secp256_test.go]: It's usually better to use the same input data for all iterations.
package testing runs benchmarks several times to make the
result more stable, so `StartTimer`, `StopTimer`  is not required.
This also applies to BenchmarkSign.

```go
_, seckey := GenerateKeyPair()
msg := randentropy.GetEntropyCSPRNG(32)
sig, _ := Sign(msg, seckey)
for i := 0; i < b.N; i++ {
    if _, err := RecoverPubkey(msg, sig); err != nil {
        b.Fatal(err)
    }
} 
````

File: crypto/secp256k1/secp256.go [hunk 2]
```diff
@@ -116,39 +119,54 @@ func GeneratePubKey(seckey []byte) ([]byte, error) {
 	return pubkey, nil
 }
 
+// NOTE: this function zeroes the input privkey. Callers need to
+// ensure they have a copy if they need to use the key after calling
+// this function.
 func Sign(msg []byte, seckey []byte) ([]byte, error) {
-	nonce := randentropy.GetEntropyCSPRNG(32)
+	defer zeroBytes(seckey)
```
Comment (fjl) [inline crypto/secp256k1/secp256.go]: I think the zeroing should be done in package crypto, where the temporary slice is allocated.

File: crypto/secp256k1/secp256_test.go (hunk 1 above)
Comment (Gustav-Simonsson) [inline crypto/secp256k1/secp256_test.go]: Actually in this benchmark it's intended to use a new key for every recovery to avoid potential caching affecting the benchmark. I'll add another benchmark where the same key is used though, that will be interesting to see if caching affects the result. In that one starting and stopping of the timer can indeed be skipped.

Comment (fjl) [inline crypto/secp256k1/secp256_test.go]: What kind of caching? libsecp256k1 shouldn't cache anything.

Comment (Gustav-Simonsson) [inline crypto/secp256k1/secp256_test.go]: I don't know, but it takes much less time if using the same key. See new commit - pushed a separate benchmark for using the same key.

**Comment by fjl** [comment]
:+1: and please squash one more time.
